### PR TITLE
[Prototype] Custom quantum gradients

### PR DESCRIPTION
**Pull request by josh146**
Quick prototype to see test how easy it is to support custom quantum gradient scripts.

Example script:

```python
import pennylane as qml
from pennylane import numpy as np


qml.enable_tape()

dev = qml.device("default.qubit", wires=2)

@qml.custom_gradient
def forward_fd(tape, params):
    """This custom quantum gradient function simply implements
    forward finite differences. It accepts the tape to differentiate and the
    parameters, and returns the quantum gradient tapes and a function
    to classically post-process the results."""
    h = 1e-7

    tape_y0 = tape.copy(copy_operations=True)
    tape_y0.set_parameters(params)
    grad_tapes = [tape_y0]

    for idx, _ in enumerate(params):
        shift = np.zeros_like(params, dtype=np.float64)
        shift[idx] = h

        tape_forward = tape.copy(copy_operations=True)
        tape_forward.set_parameters(params + shift)

        grad_tapes.append(tape_forward)

    def processing_fn(results):
        print("Hello from within my custom grad function")
        y0 = results[0]
        y_shifted = results[1:]
        return ((y_shifted - y0) / h).T

    return grad_tapes, processing_fn


@forward_fd
@qml.qnode(dev)
def circuit(weights):
    qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

weights = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)

res = circuit(weights)
print("Result:", res)

grad = qml.grad(circuit)(weights)
print("Gradient:", grad)
```

Things that should be addressed beyond a prototype:

* The `@qml.custom_gradient` decorator currently just mutates the QNode, for rapid prototyping. But this should be easy enough to turn into a pure function.

* The steps for tape manipulation are not easy nor intuitive. We should work on making tapes easy to manipulate (avoiding the setting of parameters and the copying) so that it is easy to create custom quantum gradients.

### CODE REVIEW

**Comment by mariaschuld** [comment]
Amazing!

**Comment by mariaschuld** [comment]
So in the end this would be most likely used as:
``` python
circuit = qml.custom_gradient(circuit, forward_fd)
```
right?

**Comment by josh146** [comment]
Yep, the following works as is:

```python

@qml.qnode(dev)
def circuit(weights):
    qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

weights = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)

res = circuit(weights)
print("Result:", res)

grad = qml.grad(circuit)(weights)
print("Parameter shift:", grad)

circuit2 = qml.custom_gradient(forward_fd)(circuit)
grad = qml.grad(circuit2)(weights)
print("Custom gradient:", grad)
```

Giving output:

```python
Result: 0.9653149833756958

Parameter shift: (array([[[ 1.11022302e-16,  2.39592113e-01, -3.08888780e-03],
        [ 0.00000000e+00,  1.23816921e-02, -2.69308200e-04]],

       [[-2.69308200e-04,  6.89934345e-02,  2.98265130e-04],
        [-3.35310906e-03,  2.40090367e-01,  0.00000000e+00]],

       [[ 0.00000000e+00,  0.00000000e+00,  1.11022302e-16],
        [-2.03249560e-04,  1.00153410e-01,  0.00000000e+00]]]),)

Hello from within my custom grad function
Custom gradient: (array([[[ 2.22044605e-09,  2.39592071e-01, -3.08888581e-03],
        [ 4.44089210e-09,  1.23816990e-02, -2.69304579e-04]],

       [[-2.69306799e-04,  6.89933877e-02,  2.98268077e-04],
        [-3.35310446e-03,  2.40090315e-01,  6.66133815e-09]],

       [[ 4.44089210e-09,  4.44089210e-09,  2.22044605e-09],
        [-2.03246309e-04,  1.00153367e-01,  4.44089210e-09]]]),)
```

(note that parameter-shift has better accuracy like we expect 😆 )

**Comment by josh146** [comment]
It's very flexible, even this permutation works:

```python
@qml.custom_gradient
def forward_fd(tape, params):
    h = 1e-7

    tape_y0 = tape.copy(copy_operations=True)
    tape_y0.set_parameters(params)
    grad_tapes = [tape_y0]

    for idx, _ in enumerate(params):
        shift = np.zeros_like(params, dtype=np.float64)
        shift[idx] = h

        tape_forward = tape.copy(copy_operations=True)
        tape_forward.set_parameters(params + shift)

        grad_tapes.append(tape_forward)

    def processing_fn(results):
        print("Hello from within my custom grad function")
        y0 = results[0]
        y_shifted = results[1:]
        return ((y_shifted - y0) / h).T

    return grad_tapes, processing_fn


@qml.qnode(dev)
def circuit(weights):
    qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

weights = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)

res = circuit(weights)
print("Result:", res)

# built in parameter shift
grad = qml.grad(circuit)(weights)
print("Parameter shift:", grad)

# custom gradient
circuit2 = forward_fd(circuit)
grad = qml.grad(circuit2)(weights)
print("Custom gradient:", grad)
```

**Comment by Thenerdstation** [comment]
I absolutely love this.

**Comment by Thenerdstation** [comment]
I feel so seen with this PR.

File: pennylane/tape/tapes/jacobian_tape.py [hunk 1]
```diff
@@ -523,6 +525,11 @@ def jacobian(self, device, params=None, **options):
         options["device"] = device
         options["dev_wires"] = device.wires
 
+        if custom_diff_method is not None:
+            tapes, processing_fn = custom_diff_method(self, params)
```
Comment (Thenerdstation) [inline pennylane/tape/tapes/jacobian_tape.py:529]: I love the idea of returning another high level method.

**Comment by Thenerdstation** [comment]
I see a lot of 

```python
tape_forward = tape.copy(copy_operations=True)
tape_forward.set_parameters(...)
```

We could clean this up by adding a `with_parameters` method to tape, so the code would instead look like:

```
tape_forward = tape.with_parameters(...)
```

The machinery would be the same, but it would be cleaner for the user. We then would discourage the use of `set_parameters` by renaming it to `_set_parameters`

**Comment by josh146** [comment]
> `tape_forward = tape.with_parameters(...)`

I like this! I wonder if we could avoid it being a method? Something like

```python
>>> tape_forward = qml.map(tape, parameters)
```

**Comment by Thenerdstation** [comment]
Would the name collision with `qml.map : T, P -> T` and `qml.map: [Qnode] -> QCollection` be confusing?

**Comment by Thenerdstation** [comment]
I'm also ok with `qml.with_parameters(tape, params)`.

**Comment by trbromley** [comment]
@josh146, you won the 1000th issue/PR prize! :gift:

**Pull request closed without merging**